### ISSUE TITLE
Project setup

### ISSUE THREAD

**Issue by jamesMcMeex** [edgarfinn/Tune-matching-logic#5]
Let’s start thinking about some common conventions for the project setup. Things like:

- Linting (I know, mentioned in another ticket)
- Editor config (spaces or tabs)
- Environment setup (not sure but think we should consider if this is a thing we should / need to worry about)
- Contributor guidelines (we may need some rules to govern how we interact with the repo, as we consider operating with open source conventions)

After we’ve started some exploratory work we will likely get to the point where we will need to start writing more formal documentation.

- README file (more detail on what the app is, how to get / download / install / use the app)

**Comment by edgarfinn** [comment]
Sounds good. Definitely worth at least having the conversation about linting, editor config and env setup. I'd be keen to hear you think would work best for the initial push. 

My immediate thoughts:

### Linting
I think its `eslint` that has a few good out-of-the-box style-guides. From experience, some can be too strict (ie airbnb style-guide). JS-Standard could be good? We can always add / change rules as we see fit.

### Editor config:
I cant think of anything other than spaces vs tabs (and having linters installed). Anything else you can think of?

### Environment setup:
I imagine we'll need a server each for front / back end, and might need an env file for storing ports. 
Until / unless we start using any 3rd party apis or OAuth, I cant imagine we'll need to store a whole lot of env variables, so should be fairly minimal.

### Contributor guidelines:
I'm cautious of making things so strenuous that we never get anything done, or spend too much time code reviewing. I think it depends on how much we each want to "own" different aspects of the codebase. 

- One the one hand, we might prefer to divvy up roles between front and back end, and therefore find we might not need to understand and approve of what/how the other one is authoring things. You may have no interest in SQL for example? I dont know.

- On the other hand, if we do both want to have complete ownership across the code base, we could learn more, and I do respect that approach in that either of us can work on any part of the project, which can be good for switching things up based on learning desires / day-to-day motivation.

To start with, I'd be inclined to suggest simply opening a PR for each feature / update / bugfix etc... and requiring it reviewed and approved by the other before it can be merged.

As the project grows Im sure we may want to evolve that. I'd be keen to learn about implementing some CI testing, and test coverage. But that requires that we have decent text coverage before we can consider that

**Comment by edgarfinn** [comment]
I guess another couple of possibilities on contributing are 
- we could block pushing straight to master
- we could use a kind of "staging" branch that sits between master and each feature branch. So `staging` acts as our master, we pull down latest from `staging` and rebase from it, and only use `master` to deploy, and ensure the staging branch is only ever merged into master when we're happy that its stable.

It might be overkill and cumbersome, particularly to begin with, so maybe this is for further down the line...?

**Comment by jamesMcMeex** [comment]
### eslint
Most definitely - you've put in an initial eslint config I see, let's build the config we both would like to use. Agree the Airbnb config is a bit odd IMO. I'd be happy to go with the standard one, review and extend it. It's more about keeping a consistent code authoring style, so both contributions look and feel homogenous to the app.

### editorconfig
Not much to this really, check this out: https://editorconfig.org/. I got everyone using it at work, so that individual contributors to a project didn't accidentally create huge commits full of character diffs. You can control line endings (more important for interaction between Unix and Windows contributors), and other low level file and character composition. It just preempts some headaches. For VS Code you need the editorconfig extension: https://marketplace.visualstudio.com/items?itemName=EditorConfig.EditorConfig

### Contibutor guidelines
Agree, maybe this is overkill. I also don't know if I'd actually be able to review your code, though I can QA stuff, like a UAT step I guess. I'd certainly want to know how the back end works and I hope to contribute a bit. Though I think you'll be able to appraise the front end just fine I reckon.

Code review for PRs - yes
Testing - yes. No idea how, but let's learn it. I have some idea of CI from where I work, but not got my hands dirty. Suspect it's just a lot of scripting, we can start small and build it up.

Last comment on git workflow - I like it. I think that is sensible, and though as you say potentially a bit cumbersome, I think it will help us in being responsible developers. We have a clunky workflow on the project I am working on in the day job currently, but it largely works and it keep the quality of the project consistent (but it's not perfect).

I think these subjects could be broken up into individual tickets.
